dhcpv6/server.go

```diff
@@ -4,43 +4,121 @@ import (
 	"fmt"
 	"log"
 	"net"
+	"sync"
+	"time"
 )
 
-type ResponseWriter interface {
-	LocalAddr() net.Addr
-	RemoteAddr() net.Addr
-	WriteMsg(DHCPv6) error
-	Write([]byte) (int, error)
-	Close() error
+/*
+  To use the DHCPv6 server code you have to call NewServer with two arguments:
+  - a handler function, that will be called every time a valid DHCPv6 packet is
+      received, and
+  - an address to listen on.
+
+  The handler is a function that takes as input a packet connection, that can be
+  used to reply to the client; a peer address, that identifies the client sending
+  the request, and the DHCPv6 packet itself. Just implement your custom logic in
+  the handler.
+
+  The address to listen on is used to know IP address, port and optionally the
+  scope to create and UDP6 socket to listen on for DHCPv6 traffic.
+
+  Example program:
+
+
+package main
+
+import (
+	"log"
+	"net"
+
+	"github.com/insomniacslk/dhcp/dhcpv6"
+)
+
+func handler(conn net.PacketConn, peer net.Addr, m dhcpv6.DHCPv6) {
+	// this function will just print the received DHCPv6 message, without replying
+	log.Print(m.Summary())
 }
 
-type Handler interface {
-	ServeDHCP(w ResponseWriter, m *DHCPv6)
+func main() {
+	laddr := net.UDPAddr{
+		IP:   net.ParseIP("::1"),
+		Port: 547,
+	}
+	server := dhcpv6.NewServer(laddr, handler)
+
+	defer server.Close()
+	if err := server.ActivateAndServe(); err != nil {
+		log.Panic(err)
+	}
 }
 
+*/
+
+// Handler is a type that defines the handler function to be called every time a
+// valid DHCPv6 message is received
+type Handler func(conn net.PacketConn, peer net.Addr, m DHCPv6)
+
+// Server represents a DHCPv6 server object
 type Server struct {
-	PacketConn net.PacketConn
+	conn       net.PacketConn
+	connMutex  sync.Mutex
+	shouldStop chan bool
 	Handler    Handler
+	localAddr  net.UDPAddr
+}
+
+// LocalAddr returns the local address of the listening socked, or nil if not
+// listening
+func (s *Server) LocalAddr() net.Addr {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+	if s.conn == nil {
+		return nil
+	}
+	return s.conn.LocalAddr()
 }
 
+// ActivateAndServe starts the DHCPv6 server
 func (s *Server) ActivateAndServe() error {
-	if s.PacketConn == nil {
-		return fmt.Errorf("Error: no packet connection specified")
+	s.connMutex.Lock()
+	if s.conn == nil {
+		conn, err := net.ListenUDP("udp6", &s.localAddr)
+		if err != nil {
+			return err
+		}
+		s.conn = conn
 	}
```

Comment:
Maybe put the `defer s.conn.Close()` here?

Reply:
good point. I am actually thinking that this should also set `s.conn = nil` or a subsequent call to `s.Close()` would panic on an already closed connection